hazelcast/src/jepsen/hazelcast.clj

```diff
@@ -257,55 +329,235 @@
+       )
+
+     )
+    )
+  )
+
+(defn cp-semaphore-client
+  ([] (cp-semaphore-client nil nil))
+  ([conn semaphore]
+   (reify client/Client
+     (setup! [_ test node]
+       (let [conn (connect node)
+             sem (.getSemaphore (.getCPSubsystem conn) "jepsen.cpSemaphore")
+             _ (.init sem NUMBER_OF_PERMITS)]
+         (cp-semaphore-client conn sem)
+         )
+       )
+
+     (invoke! [this test op]
+       (let [clientName (.getName conn)]
+         (try
+           (info (str " " CLIENT_ID_SEPARATOR " " clientName " " CLIENT_ID_SEPARATOR " " op))
+           (case (:f op)
+             :acquire (if (.tryAcquire semaphore 5000 TimeUnit/MILLISECONDS)
+                        (do
+                          (log-acquire-ok clientName op)
+                          (assoc op :type :ok :value {:client clientName :uid (:value op)})
+                          )
+                        (do
+                          (log-acquire-fail clientName op)
+                          (assoc op :type :fail :debug {:client clientName :uid (:value op)}))
+                        )
+             :release (do
+                        (.release semaphore)
+                        (log-release-ok clientName op)
+                        (assoc op :type :ok :value {:client clientName :uid (:value op)})
+                        )
+             )
+           (catch IllegalArgumentException e
+             (case (:f op) :acquire (log-acquire-fail clientName op) :release (log-release-fail clientName op))
+             (assoc op :type :fail :error :not-permit-owner :debug {:client clientName :uid (:value op)})
+             )
+           (catch IOException e
+             (condp re-find (.getMessage e)
+               ; This indicates that the Hazelcast client doesn't have a remote
+               ; peer available, and that the message was never sent.
+               #"Packet is not send to owner address"
+               (do
+                 (case (:f op) :acquire (log-acquire-fail clientName op) :release (log-release-fail clientName op))
+                 (assoc op :type :fail :error :client-down :debug {:client clientName :uid (:value op)})
+                 )
+               (do
+                 (case (:f op) :acquire (log-acquire-maybe clientName op e) :release (log-release-maybe clientName op e))
+                 (assoc op :type :info :error :io-exception :debug {:client clientName :uid (:value op)}))
+               )
+             )
+           (catch Exception e
+             (case (:f op) :acquire (log-acquire-maybe clientName op e) :release (log-release-maybe clientName op e))
+             (assoc op :type :info :error :exception :debug {:client clientName :uid (:value op)})
+             )
+           )
+         )
+       )
+
+     (teardown! [this test]
+       (try
+         (.terminate (.getLifecycleService conn))
+         (catch Exception e
+           (warn "exception while client termination")
+           (.printStackTrace e)
```

Comment:
This is a bit worrisome; you generally want Jepsen to abort if you can't tear clients down safely, because it might cause you to leak resources over the course of a test.